Redesign of install process and screens for improved branding and consistency.

johnalarcon opened the pull request:
This PR pertains to issue #210 and implements a redesign of the installation process and views.
It adds consistency and brand recognition to the entire process and cleans up the related code.

## Description
The goals achieved with this PR are several:
1. create a faster install process, akin to an "Amazing 60 Second Install" (or something)
2. replace coin-style logo with wordmark logo
3. brand primary install screens with logo, CP color, and CTA-style buttons
4. brand secondary (error) screens with logo and secondary-style buttons (or links, as appropriate)
5. remove verbosity from texts; link to help-docs to avoid lengthy in-place descriptions
6. improve font sizes
7. remove needless steps
8. cascade in CSS rules for button colors; avoid changing rules affecting entire admin

### About the Install Process
This update breaks the install process into several main views, each of which have a CTA-stlye button:
1. Choose language
2. Setup database
3. Setup admin
4. Enjoy success

There are also quite a few supporting views for the many various errors that might occur during installation. These are handled and branded, as well.

### About the Logo File
I've added png/svg logo files to `./wp-admin/images/` as they're install-related - perhaps these should be moved to `./wp-includes/images/` where there are other brand assets. The filenames should be reviewed and a naming schema decided.
	
### About the Texts
Most of the install and error texts have been rewritten. The goal was to remove verbosity, yet remain informative. I feel a nice balance has been struck, but am open to suggestions as few of these have been discussed. A notable exception is in the case of database errors - this markup is provided to the install process, not generated within it, so it is used as-is.
	
### Additional
Probably the biggest change code-wise is the removal of the on/off PHP tags. My syntax highlighting was ineffective with it and my IDE's code collapse was "dancing"... I hope this is not a problem. In some places, I've outputted tabs and newlines to make the source markup readable; these aren't really needed and can be removed.

### Bonus
The "Alright Sparky" step is removed in support of this [petition](https://petitions.classicpress.net/posts/95/remove-the-alright-sparky-step-during-installation). This petition has decent support, but does not reach the 50 vote threshold. As we are pre-1.0, perhaps a committee decision can be made to allow it. This step can be added back in, if absolutely necessary.

## How has this been tested?
Manual, localhost, painstakingly. That said, it will be good to get a bunch of eyes on it.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Non-breaking enhancement.

## Checklist:
- [x] My code follows the code style of this project.
- [x] It is not clear if this change will require documentation updates.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Screenshots:

### Primary Views

**Choose Language**
![main-choose-language](https://user-images.githubusercontent.com/12477319/49679002-18c03980-fa3d-11e8-8af2-b913f2e4d606.jpg)

**Database Setup**
![main-database-setup](https://user-images.githubusercontent.com/12477319/49679042-6046c580-fa3d-11e8-84c7-c68b3f6925ae.jpg)

**Admin Setup**
![main-admin-setup](https://user-images.githubusercontent.com/12477319/49679054-705ea500-fa3d-11e8-8f8f-505ae27bef06.jpg)

**Install Success** when user chose own password
![main-success-1](https://user-images.githubusercontent.com/12477319/49679071-92582780-fa3d-11e8-9798-7475c6adc41b.jpg)

**Install Success** when system chose the password
![main-success-2](https://user-images.githubusercontent.com/12477319/49679081-984e0880-fa3d-11e8-9492-2a480e51024b.jpg)


### Secondary Views

**Warning** config file exists in install
![warning-config-file-found-1](https://user-images.githubusercontent.com/12477319/49679171-4bb6fd00-fa3e-11e8-87f0-abb1fc58b70e.jpg)

**Warning** config file exists 1 level up from install
![warning-config-file-found-2](https://user-images.githubusercontent.com/12477319/49679235-c253fa80-fa3e-11e8-9b7a-853052530e1b.jpg)

**Error** selecting database
![error-database-selection](https://user-images.githubusercontent.com/12477319/49679241-ca139f00-fa3e-11e8-8168-71dc4ca35659.jpg)

**Error** connecting to database
![error-database-connection](https://user-images.githubusercontent.com/12477319/49679248-d13aad00-fa3e-11e8-90fa-cad07e3342f1.jpg)

**Error** missing table prefix
![error-table-prefix-missing](https://user-images.githubusercontent.com/12477319/49679252-dbf54200-fa3e-11e8-99da-fbe87cd4d655.jpg)

**Error** invalid table prefix
![error-table-prefix-invalid](https://user-images.githubusercontent.com/12477319/49679254-e0215f80-fa3e-11e8-842a-e9ab784c697f.jpg)

**Error** invalid username
![error-admin-username](https://user-images.githubusercontent.com/12477319/49679257-eb748b00-fa3e-11e8-924c-1d4b0e368c32.jpg)

**Error** invalid email
![error-admin-email](https://user-images.githubusercontent.com/12477319/49679269-fb8c6a80-fa3e-11e8-84fc-de4c5a674fe4.jpg)

**Error** insufficient MySQL version
![error-mysql-insufficient](https://user-images.githubusercontent.com/12477319/49679278-08a95980-fa3f-11e8-8a00-cfda21057ce9.jpg)

**Error** insufficient PHP version
![error-php-insufficient](https://user-images.githubusercontent.com/12477319/49679281-0cd57700-fa3f-11e8-8e27-fa053f762d83.jpg)

**Error** insufficient PHP and MySQL version
![error-php-mysql-insufficient](https://user-images.githubusercontent.com/12477319/49679283-13fc8500-fa3f-11e8-84b4-308347d5ca68.jpg)

**Error** config file permissions
![error-config-file-permissions](https://user-images.githubusercontent.com/12477319/49679286-1ced5680-fa3f-11e8-870b-149da52075d4.jpg)

**Error** empty table prefix
![error-table-prefix-empty](https://user-images.githubusercontent.com/12477319/49679296-270f5500-fa3f-11e8-9dd6-863bdf7f653d.jpg)

**Error** preventative constant set
![error-prevented-by-constant](https://user-images.githubusercontent.com/12477319/49679304-342c4400-fa3f-11e8-82be-8c0bbc791458.jpg)

**Error** already installed
![error-already-installed](https://user-images.githubusercontent.com/12477319/49679309-41493300-fa3f-11e8-8139-fb49ff12c930.jpg)

**Error** sample config file missing
![error-missing-sample-config](https://user-images.githubusercontent.com/12477319/49679320-4dcd8b80-fa3f-11e8-9674-2e19041da086.jpg)

There are more views, as well, such as when the user tries to add wonky characters to the table prefix... or uses unsupported characters in their email address. Those are also accounted for. I've just run short on time and need to get this posted.

octoxan commented:
Looks fantastic!

nylen commented on src/wp-admin/install.php:
```diff
@@ -229,11 +66,13 @@ function display_setup_form( $error = null ) {
+// Insufficient PHP and/or MySQL? Set a flag with an error message.
 if ( !$mysql_compat && !$php_compat ) {
 	/* translators: 1: ClassicPress version number, 2: Minimum required PHP version number, 3: Minimum required MySQL version number, 4: Current PHP version number, 5: Current MySQL version number */
 	$compat = sprintf( __( 'You cannot install because <a href="https://codex.wordpress.org/Version_%1$s">ClassicPress %1$s</a> requires PHP version %2$s or higher and MySQL version %3$s or higher. You are running PHP version %4$s and MySQL version %5$s.' ), $wp_version, $required_php_version, $required_mysql_version, $php_version, $mysql_version );
```
Let's update these strings too.  Right now they indicate the WordPress-equivalent version (currently 4.9.8) and link to the WP codex.

The logic for determining the link to the ClassicPress version can work as follows:

- If `classicpress_version()` does not contain a `+` character, link to `https://github.com/ClassicPress/ClassicPress-release/releases/%1$s` (e.g. https://github.com/ClassicPress/ClassicPress-release/releases/1.0.0-beta1).
- If `classicpress_version()` contains a `+` character, there are several possibilities for why, not all of which actually have an official release build.  In that case we should probably just link to the list at https://github.com/ClassicPress/ClassicPress-release instead.

johnalarcon replied:
👍

johnalarcon replied:
Updated.

nylen commented on src/wp-admin/install.php:
```diff
@@ -245,34 +84,36 @@ function display_setup_form( $error = null ) {
-		) . '</p></body></html>'
-	);
+	display_header( 'cp-installation-error' );
+	echo '<h1>'. __('Configuration Error') . '</h1>';
```
Can add spaces here `__( 'Configuration Error' )`.

johnalarcon replied:
Added.

nylen commented on src/wp-admin/install.php:
```diff
@@ -299,119 +147,286 @@ function display_setup_form( $error = null ) {
+		// Print the branded page header.
 		display_header();
-		// Fill in the data we gathered
+		// Get and pare submitted data.
```
`parse` instead of `pare`?

johnalarcon replied:
Pare...to trim. I've reworded it for clarity.

nylen commented on src/wp-admin/install.php:
```diff
@@ -299,119 +147,286 @@ function display_setup_form( $error = null ) {
+			$error = __( 'Sorry, that isn&#8217;t a valid email address. Email addresses look like <code>username@example.com</code>.' );
 		}
-
+		if ($error) {
```
Can add spaces here: `if ( $error )`.

johnalarcon replied:
Added.

nylen commented on src/wp-admin/install.php:
```diff
@@ -299,119 +147,286 @@ function display_setup_form( $error = null ) {
+		echo '			<input type="checkbox" name="pw_weak" class="pw-checkbox" />' . __( 'Confirm use of weak password' );
+		echo '			</label>' . "\n";
+		echo '		</td>' . "\n";
+		echo '	</tr>' . "\n";
```
I like this approach to PHP/HTML soup.  The `echo` statements and the indentation provide visual alignment in the source code, and the `"\n"` makes the HTML more readable after it gets to the browser.  It's often necessary to inspect and make sure this is working as intended, so I think it's worth including.

johnalarcon replied:
Great! I was dreading the possibility of a request to revert it back. I'm not a fan of a ton of echos, but the points you make are exactly why I took this approach. 👍

nylen commented:
Excellent work @johnalarcon!  I will test this out more thoroughly, but for now I've left a few minor comments on the code.

The build failure is because a new image was added without being minified:

![2018-12-08t11 47 51-05 00](https://user-images.githubusercontent.com/227022/49688395-46f65580-fadf-11e8-9a9f-78f5f12efe4d.png)

If you can run `npm install` and `grunt precommit` that will minify the images.  Then you should be able to commit the changes.

This can also wait until we organize the images a bit better.

I will give this some more thought, but for now, one consideration is that we need to keep the filenames of the old WordPress logo images the same, so that plugins/themes/sites can reference them the same way.

This means we either need to fit new images into the existing naming scheme, or make copies of any new images.  I'm not sure which is the best way, but so far we have been fitting things in as best we can.

johnalarcon commented:
> Excellent work @johnalarcon! I will test this out more thoroughly, but for now I've left a few minor comments on the code.

Thanks! I've addressed some of the comments and updated the PR accordingly. I'll get to the remaining items soon.

> The build failure is because a new image was added without being minified:
The irony...

> If you can run `npm install` and `grunt precommit` that will minify the images. Then you should be able to commit the changes.

Ok.

> This can also wait until we organize the images a bit better.

> I will give this some more thought, but for now, one consideration is that we need to keep the filenames of the old WordPress logo images the same, so that plugins/themes/sites can reference them the same way.

> This means we either need to fit new images into the existing naming scheme, or make copies of any new images. I'm not sure which is the best way, but so far we have been fitting things in as best we can.

I've created issue #266 with a proposal to get this in order.

nylen commented:
From https://www.classicpress.net/democracy/#democracy-exceptions:

> As you can imagine, if every decision required a Petition or Referendum, the community would be bogged down with an unmanageable bureaucratic process that would stifle progress. As a result, there are a number of processes and tasks that don’t need to go through the Petition or Referendum process. These include:
>
> Minor modifications to existing functionality or processes

I think removing the Alright Sparky step qualifies, and it's a clear improvement with (as far as I know) no backwards compatibility implications.

It's also already implemented, so even better.

The code looks good at a glance, but I agree this needs a thorough review.  I haven't forgotten about it, and we'll get to it as soon as we can.

johnalarcon commented:
Thanks for the update.

> > Minor modifications to existing functionality or processes
> 
> I think removing the Alright Sparky step qualifies, and it's a clear improvement with (as far as I know) no backwards compatibility implications.

Great!

> The code looks good at a glance, but I agree this needs a thorough review. I haven't forgotten about it, and we'll get to it as soon as we can.

No worries.

nylen commented on src/wp-admin/images/cp-logo-gradient.svg:
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 238.59 46.29"><defs><linearGradient id="WCAG_compliant_Aqua" x1="220.41" y1="28.1" x2="206.08" y2="13.77" gradientUnits="userSpaceOnUse"><stop offset="0" stop-color="#3dbca6"/><stop offset="0.5" stop-color="#017f98"/><stop offset="1" stop-color="#006b81"/></linearGradie ...
```
Please use this SVG file directly and remove the `.png` added here (it's very low-resolution anyway).  Fortunately, we are past the point where we need to care about browsers which can't handle SVG images, and we've already added several others.

johnalarcon replied:
I've updated the style sheet to use the direct URL for the logo image, but am unable to delete the 2 images from the pull request...I get a non-working delete button for the svg, and no delete button for the png.

nylen replied:
The GitHub web UI doesn't support this, I guess.  I can take care of it.

johnalarcon replied:
Thanks. I'll let you mark this as resolved when ready.

nylen replied:
Done.  Note, I updated the stylesheet to use the copy of the SVG from the core code instead of a remote URL.

johnalarcon commented:
After reviewing Beta2, I notice the initial step of the installation – the language picker – is showing a lot of white space due to there being fewer languages. I think there were 11 of them. It's not clear when there will be additional languages added, so I was thinking it should be reduced in height to fit, so it doesn't give the impression that something is missing. What say you?

nylen commented:
> I was thinking [the language picker in the installation] should be reduced in height to fit, so it doesn't give the impression that something is missing. What say you?

Sounds good to me.

johnalarcon commented:
I changed the size a few days back... forgot to drop a comment.

invisnet commented on src/wp-admin/install.php:
```diff
@@ -299,119 +170,286 @@ function display_setup_form( $error = null ) {
+			echo '<table class="form-table install-success">' . "\n";
+			echo '<tr>' . "\n";
+			echo '<th>' . __( 'Username' ) . '</th>' . "\n";
+			echo '<td><code>' . esc_html( sanitize_user( $user_name, true ) ) . '</code></td>' . "\n";
```
Why do we need to sanitize the user at this point?

nylen replied:
We've already verified that the user is sanitized.  I've removed this call.

nylen commented:
Thanks for doing this @johnalarcon, it took a while but we are finishing it.

There were more button styles and colors here than you might expect, so I've modified them to look like the buttons in #223:

![2019-02-19t02 27 07-05 00](https://user-images.githubusercontent.com/227022/52997240-f0b26680-33ed-11e9-89e1-ba277b56cb54.png)

This darker color scheme also has a better contrast ratio, and I've added some fixes for disabled buttons and other edge cases.

I was able to test this including all the progress / error / secondary screens.  Tomorrow I'll merge #366 first and then update this PR to use the new filename.

johnalarcon commented:
Sounds good; thanks for the update.
